editors/code/src/inlay_hints.ts

```diff
@@ -3,20 +3,21 @@ import * as vscode from 'vscode';
 import * as ra from './rust-analyzer-api';
 
 import { Ctx, Disposable } from './ctx';
-import { sendRequestWithRetry, isRustDocument, RustDocument, RustEditor } from './util';
+import { sendRequestWithRetry, isRustDocument, RustDocument, RustEditor, sleep } from './util';
 
 
 export function activateInlayHints(ctx: Ctx) {
     const maybeUpdater = {
         updater: null as null | HintsUpdater,
-        onConfigChange() {
+        async onConfigChange() {
             if (
                 !ctx.config.inlayHints.typeHints &&
                 !ctx.config.inlayHints.parameterHints &&
                 !ctx.config.inlayHints.chainingHints
             ) {
                 return this.dispose();
             }
+            await sleep(10);
```

Comment:
let's bump this to 100, as this depends not only on the event loop in the extension itself, but on the event loop in the server. But also, urgh, causality broken once again!

Reply:
(caution: noob rust-analyzer user, excuse dumb questions for now!)
Is it possible to have the server keep causality in case of config update event specifically?
As in, everything sent after config update event, executes after the event? 
Perhaps the server can block all tasks on receiving config update notification, to ensure no task can "race" config update.
Otherwise I feel this issue of broken causality will happen in other parts of configuration updates, not just inlay_hints.ts.

Reply:
Actually, I think we can handle this correctly without major changes. 

We should add a middlewhere for `getConfigurationRequest`, and schedule an update after we've responded.

Probably need a `didChangeServerConfig` event for this

Reply:
@matklad I've tried the middleware approach and it failed, because the middleware returns the response to the server, but you never know when this response is delivered and when the server has applied the update

Reply:
Actaully, blocking while the config update is in progress as proposed by @nokola seems like an approach that guarantees consistency, since the server will never use the out-of-date config resonding for inlay hints.

This whole problem appeared because we did our custom implementation of inlay hints, it might appear if we implement anything custom like inlay hints in future

Reply:
I feel strongly that the server should just process requests and responses in-order, as they arrive. This *could* be enough to ensure causality, if the client cooperates. 

Holding of responses is much harder mental model, and it deadlocks in the workstace, where the underlying request queue is filled.

Reply:
@matklad So there might be the problem with the order of processing on the server, because the middleware approach doesn't work for me, the inlay hints request somehow is processed before the config update response. I may have done some mistake, let me double check and maybe create another PR for that approach

Reply:
@matklad what do you mean by “holding of responses” mental model?

Reply:
Never mind, I re-read the thread, I assume you meant "server does not start another request until current one is finished".
I assume the server _starts_ processing requests in order but does not _finish_ processing requests in order.